# Prow Logs and GCS Data

In [1]:
import gzip
import json
import requests
import pandas as pd
from bs4 import BeautifulSoup

### What do we have access to as data scientists when digging into the build artifacts?

In this notebook we want to determine how to interact with the availble data on GCS. We will provide some EDA and perfrom 2 different tasks.


1. Compare the different log files present throughout the archives and quantify how complete and compareable our log dataset is.
1. Download a small sample dataset of events and build logs to perform some EDA on.  


In [2]:
# Load test file to access our record of job names
with gzip.open("../../../data/raw/testgrid_810.json.gz", "rb") as read_file:
    testgrid_data = json.load(read_file)

### Example to access a single prow artifacts page

Let's make sure we understand how this works, and focus on a single job first.

In [3]:
tab = '"redhat-openshift-ocp-release-4.6-informing"'
grid = "periodic-ci-openshift-release-master-ci-4.6-upgrade-from-stable-4.5-e2e-gcp-upgrade"

In [4]:
response = requests.get(
    f"https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/logs/{grid}"
)
soup = BeautifulSoup(response.text, "html.parser")
list_of_builds = [x.get_text()[1:-1] for x in soup.find_all("a")][1:-1]

In [5]:
response = requests.get(
    f"https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/logs/{grid}/{list_of_builds[1]}"
)
response.url

'https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/logs/periodic-ci-openshift-release-master-ci-4.6-upgrade-from-stable-4.5-e2e-gcp-upgrade/1364869749170769920'

In [6]:
soup = BeautifulSoup(response.text, "html.parser")

In [7]:
[x.get_text() for x in soup.find_all("a")]

[' ..',
 ' artifacts/',
 ' build-log.txt',
 ' finished.json',
 ' podinfo.json',
 ' prowjob.json',
 ' started.json']

Great, we can now programmatically access the archives. Now, lets walk through all of the job archives for a single job and create a list of what they have on the first depth level of thier directory.  

In [8]:
build_data = {}

for build in list_of_builds:
    response = requests.get(
        f"https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/logs/{grid}/{build}"
    )
    soup = BeautifulSoup(response.text, "html.parser")
    artifacts = [x.get_text() for x in soup.find_all("a")]
    build_data[build] = artifacts

In [9]:
builds_info = pd.Series({k: len(v) for (k, v) in build_data.items()})

In [10]:
builds_info.value_counts()

7    139
6      3
5      1
dtype: int64

In [11]:
pd.Series(build_data).apply(" ".join).value_counts()

 ..  artifacts/  build-log.txt  finished.json  podinfo.json  prowjob.json  started.json    139
 ..  artifacts/  build-log.txt  finished.json  prowjob.json  started.json                    2
 ..  build-log.txt  finished.json  prowjob.json  started.json                                1
 ..  build-log.txt  finished.json  podinfo.json  prowjob.json  started.json                  1
dtype: int64

In [12]:
builds_info.value_counts() / len(builds_info)

7    0.972028
6    0.020979
5    0.006993
dtype: float64

97% percent of our records for this job include appear to be complete and include the 'artifacts/' subdirectory, lets dig in and see what they contain. 

In [13]:
build_data = {}

for build in list_of_builds:
    response = requests.get(
        f"https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/logs/{grid}/{build}/artifacts"
    )
    soup = BeautifulSoup(response.text, "html.parser")
    artifacts = [x.get_text() for x in soup.find_all("a")]
    build_data[build] = artifacts

In [14]:
artifact_info = pd.Series({k: len(v) for (k, v) in build_data.items()})
artifact_info.value_counts()

6    92
5    36
7     8
1     2
3     2
4     2
2     1
dtype: int64

In [15]:
artifact_info.value_counts() / len(artifact_info)

6    0.643357
5    0.251748
7    0.055944
1    0.013986
3    0.013986
4    0.013986
2    0.006993
dtype: float64

In [16]:
pd.Series(build_data).apply(" ".join).value_counts()

 ..  build-resources/  e2e-gcp-upgrade/  release/  junit_operator.xml  metadata.json                     92
 ..  build-resources/  e2e-gcp-upgrade/  junit_operator.xml  metadata.json                               27
 ..  build-resources/  release/  junit_operator.xml  metadata.json                                        9
 ..  build-resources/  e2e-gcp-upgrade/  release/  ci-operator.log  junit_operator.xml  metadata.json     8
 ..                                                                                                       2
 ..  build-resources/  junit_operator.xml  metadata.json                                                  2
 ..  junit_job.xml  metadata.json                                                                         1
 ..  e2e-gcp-upgrade/  release/                                                                           1
 ..  junit_job.xml                                                                                        1
dtype: int64

We can see that once we get down into the artifacts there is a fair bit less uniformity to the data available to us. And this is all within a single job! We will look into it next, but my assumptions is that this issue gets worse when comparing available artifacts across jobs. 

This will make it somewhat difficult to use these sets of documents as a whole to compare different CI behaviour. At this point, it might make sense to consider looking only at the same document (log) across job when availble. 

In the next section we are going to walkthrough how to access and parse the build-log as well as the events json both to download a small testing dataset our working directly with the data in memory.  

## Collecting build logs

In [17]:
build_log_data = {}
for build in list_of_builds:
    response = requests.get(
        f"https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/logs/{grid}/{build}/build-log.txt"
    )
    soup = BeautifulSoup(response.text, "html.parser")
    build_log_data[build] = soup.get_text()

In [18]:
def not_exists(x):
    """
    Checks if the data is invalid.
    """
    return x.startswith("\n\n\n\n\n\n\nGCS browser")


def get_counts(x):
    """
    Gets counts for chars, words, lines for a log.
    """
    if not not_exists(x):
        chars = len(x)
        words = len(x.split())
        lines = x.count("\n") + 1
        return chars, words, lines
    else:
        print("no logs")
        return 0, 0, 0

In [19]:
## Create a dataframe with char, words, and lines
## count for the logs
data = []
for key, value in build_log_data.items():
    chars, words, lines = get_counts(value)
    data.append([key, chars, words, lines])

df = pd.DataFrame(data=data, columns=["build_log_id", "chars", "words", "lines"])
df

build_log_id  chars  words  lines
0    1364778930069835776    517     51      5
1    1364869749170769920  10805    919    111
2    1364960659313266688  10807    919    111
3    1365051265100288000  15758   1316    184
4    1365142142841786368  10804    919    111
..                   ...    ...    ...    ...
138  1377248782869073920  11416    926    113
139  1377339577248780288  13933    926    113
140  1377430351407747072  13933    926    113
141  1377521126032281600  13932    926    113
142  1377611908198174720  13932    926    113

[143 rows x 4 columns]

#### See the stats for chars, words, lines

In [20]:
df["chars"].describe()

count    1.430000e+02
mean     1.397021e+05
std      7.443312e+05
min      9.200000e+01
25%      1.093800e+04
50%      1.095000e+04
75%      1.611400e+04
max      6.662598e+06
Name: chars, dtype: float64

In [21]:
df["words"].describe()

count       143.000000
mean      10101.440559
std       54292.735249
min          11.000000
25%         925.000000
50%         926.000000
75%        1389.500000
max      508960.000000
Name: words, dtype: float64

In [22]:
df["lines"].describe()

count      143.000000
mean       649.706294
std       2895.837016
min          2.000000
25%        112.000000
50%        112.000000
75%        171.000000
max      21291.000000
Name: lines, dtype: float64

* From the initial analysis, we see that we have log files with 2 lines to ~21,000 lines with a mean of 649 lines. This suggests high variability. The next few things to look at would be similarity betwen the log files, performing word analysis, templating, and clustering.

## Collect the events data

In [23]:
build_events_data = {}
for build in list_of_builds:
    response = requests.get(
        f"https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/\
        logs/{grid}/{build}/artifacts/build-resources/events.json"
    )
    soup = BeautifulSoup(response.text, "html.parser")
    soup_text = soup.get_text()
    if not not_exists(soup_text):
        build_events_data[build] = json.loads(soup_text)
    else:
        build_events_data[build] = {}

In [24]:
## Percentage of builds that have the events.json file
count = 0
for key, value in build_events_data.items():
    if value:
        count += 1
count * 100 / len(build_events_data)

96.5034965034965

In [25]:
# Analyzing the messages of a single build
messages = [
    (i["metadata"]["uid"], i["message"])
    for i in build_events_data["1364869749170769920"]["items"]
]
messages_df = pd.DataFrame(messages, columns=["UID", "message"])
messages_df

UID  \
0    504b881c-9e97-46a0-b206-765c9973e1d3   
1    3a92467e-993e-43a1-8eee-24ba7a22508f   
2    ed40875c-b182-4164-b730-a3754ed94124   
3    b8c5b026-936e-4f28-a0de-62051ff378d8   
4    b1b8265a-d87d-430d-b38e-c10c7f3fb91c   
..                                    ...   
477  db97f7b5-5fb4-4e68-aa82-cd9c120b9c8d   
478  947a5c3f-f318-42d8-88c5-133f9783e94d   
479  e94e0988-c2ea-4045-bb84-251410018d94   
480  00bef865-7965-4b65-b39a-9993347c5942   
481  a85f8fbd-9cac-4e47-a5d4-5fbe145d4050   

                                               message  
0    Running job periodic-ci-openshift-release-mast...  
1    Running job periodic-ci-openshift-release-mast...  
2    Running job periodic-ci-openshift-release-mast...  
3    Running job periodic-ci-openshift-release-mast...  
4                               No matching pods found  
..                                                 ...  
477  Container image "gcr.io/k8s-prow/sidecar:v2021...  
478                          Created container sidecar  
479                          Started container sidecar  
480  Back-off pulling image "image-registry.openshi...  
481                            Error: ImagePullBackOff  

[482 rows x 2 columns]

In [26]:
messages_df["message"].describe()

count                                                   482
unique                                                  156
top       Container image "gcr.io/k8s-prow/sidecar:v2021...
freq                                                     29
Name: message, dtype: object

In [27]:
messages_df["message"].value_counts().reset_index()

index  message
0    Container image "gcr.io/k8s-prow/sidecar:v2021...       29
1    Container image "gcr.io/k8s-prow/entrypoint:v2...       29
2                   Created container place-entrypoint       29
3                            Started container sidecar       29
4                            Created container sidecar       29
..                                                 ...      ...
151  Successfully pulled image "registry.ci.openshi...        1
152                         Add eth0 [10.128.34.73/23]        1
153  Successfully assigned ci-op-ft9klqc6/e2e-gcp-u...        1
154  Failed to pull image "image-registry.openshift...        1
155  Successfully pulled image "image-registry.open...        1

[156 rows x 2 columns]

In the build data, we saw that about 96.5% builds have the events.json file. We further analyzed all the events that happened for a particular build and found the frequencies of the messages. We can repeat the process for all the other builds and find most common messages and perform further analysis.

## Conclusion

In this notebook, we saw how to programmatically access the archives, build logs, and the events json. Most of the files in root directory were the same but we saw a higher variability in the contents of the artifcats directory. For the logs, we saw a high variability in the number of lines indicating that we need a more thorough EDA. The events.json lead to some neat messages that can be explored furthur for a larger dataset.